### DISCUSSION TITLE
Schema references with $ref - Difference between Draft 5 and Draft 6

### DISCUSSION THREAD

**Discussion by char0n** [json-schema-org/community#526] [Specification]
Hi everybody,

I'm seeking clarity between `Schema references with $ref`. 

For [Draft 5](https://json-schema.org/draft-05/draft-wright-json-schema-00.pdf) the definition is as follows:

> Any time a subschema is expected, a schema may instead use an object
 containing a "$ref" property. The value of the $ref is a URI
 Reference. Resolved against the current URI base, it identifies the
 URI of a schema to use. All other properties in a "$ref" object MUST
 be ignored.

For [Draft 6](https://json-schema.org/draft-06/draft-wright-json-schema-01#rfc.section.8) the definition is as follows:
> The "$ref" keyword is used to reference a schema, and provides the ability to validate recursive structures through self-reference.

> An object schema with a "$ref" property MUST be interpreted as a "$ref" reference. The value of the "$ref" property MUST be a URI Reference. Resolved against the current URI base, it identifies the URI of a schema to use. All other properties in a "$ref" object MUST be ignored.

---

From Draft 5 I can clearly infer that `{ $ref: <URI> }` is not a schema object, but rather just an object with a `$ref` property - similar with semantics to the object defined for `properties` keyword.

From Draft 6 I would infer that `{ $ref <URI> }` is a schema object with a `$ref` keyword. Also in Draft 6 I don't understand the following sentence: `An object schema with a "$ref" property MUST be interpreted as a "$ref" reference`, but again I assume it tries to clarify that the object with a `$ref` keyword is a schema object.

Would the authors please clarify the difference?

**Comment by jdesrosiers** [comment]
There was no intended behavior change for references between these drafts. I personally believe the previous wording was better and draft-04 was even better in that it just referenced a separate and reusable JSON Reference spec. It was around the time draft-05/6 were being worked on that the idea of JSON Reference being a generic re-usable thing was abandoned and started transitioning into a JSON Schema specific concept. I think that's unfortunate. People started to think about references as a special case of a schema rather than a separate concept. That's where the language change comes from. Personally, I still prefer to think of a reference as not a schema. I don't even consider it an object. It's a representation of a "reference" that happens to have similar syntax to an object.

**Comment by char0n** [comment]
Hi @jdesrosiers, 

Thanks for clarifying. Though I still have couple of questions to ask to help me fully understand the intentions ;]

> there was no intended behavior change for references between these drafts. I personally believe the previous wording was better and draft-04 was even better in that it just referenced a separate and reusable JSON Reference spec. It was around the time draft-05/6 were being worked on that the idea of JSON Reference being a generic re-usable thing was abandoned and started transitioning into a JSON Schema specific concept.

This would mean that in all prior drafts to  **Draft 2019-09**, whenever we see `{ $ref: <URI> }` we can interpret it as `JSON Reference`. This also means that `JSON Reference` must comply with https://datatracker.ietf.org/doc/html/draft-pbryan-zyp-json-ref-03 spec. This limits `JSON Reference` to Object shape with one single property of `$ref`. 

JSON Schema Example of `Draft 4/5/6/7`:

```json
{
  "type": "object",
  "properties": {
    "prop1": { "$ref": "https://exmaple.com/schema.json" }
  } 
}
```

Symbolic expression representation

```
(JSONSchema
  (Member
    (String)
    (String))
  (Member
    (String)
    (Object
      (Member
        (String)
        (JSONReference)))))     
```

---

In **Draft 2019-09**, whenever we see `{ $ref: <URI> }` it will always be interpreted as `JSON Schema with $ref keyword`. This also means, additional keywords can be used along with `$ref` keywords and the keywords from referencing and referenced JSON Schemas are merged using shallow merge right algorithm. There is no [JSON Reference](https://datatracker.ietf.org/doc/html/draft-pbryan-zyp-json-ref-03) spec involved anymore. 

This understanding would contrast with your explanation (mentioned below) but I assume you were referring to draft prior to Draft 2019-09.

> Personally, I still prefer to think of a reference as not a schema. I don't even consider it an object. It's a representation of a "reference" that happens to have similar syntax to an object

JSON Schema Example `>= Draft 2019-09`:

```json
{
  "type": "object",
  "properties": {
    "prop1": { "$ref": "https://exmaple.com/schema.json" }
  } 
}
```

Symbolic expression representation

```
(JSONSchema
  (Member
    (String)
    (String))
  (Member
    (String)
    (Object
      (Member
        (String)
        (JSONSchema)))))     
```

**Reply by jdesrosiers** [reply]
> This also means that `JSON Reference` must comply with https://datatracker.ietf.org/doc/html/draft-pbryan-zyp-json-ref-03 spec.

Technically, no, but effectively, yes. Since draft-05, the spec doesn't reference the JSON Reference spec anymore. That spec was merged into the JSON Schema spec. In that merge, nothing changed other than that it wasn't called a "JSON Reference" anymore. That remained the case until 2019-09. Since the behavior didn't change, you can still think of it as complying with the JSON Reference spec even though that's not technically what the spec says. It works because the JSON Reference spec and the JSON Schema spec never disagree.

> This limits `JSON Reference` to Object shape with one single property of `$ref`.

Correct.

> In **Draft 2019-09**, whenever we see `{ $ref: <URI> }` it will always be interpreted as `JSON Schema with $ref keyword`. This also means, additional keywords can be used along with `$ref` keywords and the keywords from referencing and referenced JSON Schemas are merged using shallow merge right algorithm.

Yes, additional keywords can be used alongside `$ref`, but, no, it's not a merge. It applies _in addition to_ any sibling keywords. Think of it like an `allOf` that only applies one schema. The new `{ "$ref": "..." }` is equivalent to the old `"allOf": [{ "$ref": "..." }]`.

The way I think about it is that the representation of a JSON Reference has changed from an object (`{ "$ref": "..." }`) to just a string (`"..."`) and the value of the `$ref` keyword is actually a reference, not a string. So, the Symbolic expression representation I would give for that last schema would be like this,

```
(JSONSchema
  (Member
    (String)
    (String))
  (Member
    (String)
    (Object
      (Member
        (String)
        (JSONSchema
          (Member
            (String)
            (JSONReference)))))))
```

If it helps, I like to imagine references using a made up syntax to remind that a reference is not a normal JSON type.

draft-04/6/7
```json
{
  "type": "object",
  "properties": {
    "prop1": < "$ref": "https://exmaple.com/schema.json" >
  } 
}
```

2029-09,2020-12
```json
{
  "type": "object",
  "properties": {
    "prop1": { "$ref": <https://exmaple.com/schema.json> }
  } 
}
```

**Reply by char0n** [reply]
@jdesrosiers thanks for all the clarifications.

### Summary

Just to sum it all up for me and others...

As you've already mentioned before: 

> There was no intended behavior change for references between these drafts (Draft 4/5/6/7)

... and even though Draft 6 and Draft 7 started to talk about `$ref` as a keyword, they further specify following:

> An object schema with a "$ref" property MUST be interpreted as a "$ref" reference

So in the light of that and all your clarifications, the following is true

---

**JSON Schema Draft 4/5/6/7**
```json
{
  "type": "object",
  "properties": {
    "prop1": { "$ref": "https://exmaple.com/schema.json" }
  } 
}
```
**Symbolic expression representation:**
```
(JSONSchema
  (Member
    (String)
    (String))
  (Member
    (String)
    (Object
      (Member
        (String)
        (JSONReference
          (Member
            (String)
            (String)))))))
```

---

**>= JSON Schema Draft 2019-09**
```json
{
  "type": "object",
  "properties": {
    "prop1": { "$ref": "https://exmaple.com/schema.json" }
  } 
}
```

**Symbolic expression representation:**
```
(JSONSchema
  (Member
    (String)
    (String))
  (Member
    (String)
    (Object
      (Member
        (String)
        (JSONSchema
          (Member
            (String)
            (JSONReference)))))))
```
> NOTE: `JSONReference` here is a semantic representation of simple string in format of URI

**Comment by char0n** [comment]
I' would also like to clarify on the following:

> Yes, additional keywords can be used alongside $ref, but, no, it's not a merge. It applies in addition to any sibling keywords. Think of it like an allOf that only applies one schema. The new { "$ref": "..." } is equivalent to the old "allOf": [{ "$ref": "..." }].

Our implementation is currently guided with https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-00#rfc.section.7.7.1.1

Which means that our dereferencing (reference removal) works like simple shallow right merge  (definition authors and implementation aggreed on this):

**Input JSON Schema:**
```json
{
    "title": "Feature list",
    "type": "array",
    "prefixItems": [
        {
            "title": "Feature A",
            "properties": {
                "enabled": {
                    "$ref": "#/$defs/enabledToggle",
                    "default": true
                }
            }
        },
        {
            "title": "Feature B",
            "properties": {
                "enabled": {
                    "description": "Feature B description",
                    "$ref": "#/$defs/enabledToggle"
                }
            }
        }
    ],
    "$defs": {
        "enabledToggle": {
            "title": "Enabled",
            "description": "generic description",
            "type": ["boolean", "null"],
            "default": null
        }
    }
}
```

**Dereferenced JSON Schema:**
```json
{
    "title": "Feature list",
    "type": "array",
    "prefixItems": [
        {
            "title": "Feature A",
            "properties": {
                "enabled": {
                    "title": "Enabled",
                    "description": "generic description",
                    "type": ["boolean", "null"],
                    "default": true
                }
            }
        },
        {
            "title": "Feature B",
            "properties": {
                "enabled": {
                    "title": "Enabled",
                    "description": "Feature B description",
                    "type": ["boolean", "null"],
                    "default": null
                }
            }
        }
    ],
    "$defs": {
        "enabledToggle": {
            "title": "Enabled",
            "description": "generic description",
            "type": ["boolean", "null"],
            "default": null
        }
    }
}
```

@jdesrosiers would you please describe on my example how the  result of dereferencing in context of your clarification would look like? Thanks a lot!

**Reply by Relequestual** [reply]
Removal of all references is not something supported by JSON Schema. There are some cases where it's not possible, and many cases where it will result in a change of behaviour and validation result.

Why do you want to remove all references? Is it due to downstream tooling?
We added bundling to the specification to make handling external references easier, but you cannot and should not try to remove all references.

Any tooling that requires removal of references is unable to fully support JSON Schema 2019-09 or above.

**Reply by jdesrosiers** [reply]
As Ben said, it's not recommended to dereference a schema this way, but if you insist, you'd have to use the `allOf` transformation I showed earlier.

```json
"enabled": {
  "allOf": [{
    "title": "Enabled",
    "description": "generic description",
    "type": ["boolean", "null"],
    "default": null
  }],
  "description": "Feature B description",
}
```

> Our implementation is currently guided with https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-00#rfc.section.7.7.1.1

That section isn't intended to allow implementations to merge schemas. It's about interpreting annotation results. The annotation results for the example above includes two `description` annotations for the "enabled" property. Both annotations must be present in the evaluation results. Section 7.7.1.1 is saying that implementations that use those annotation results may decide to pick one for their purposes rather than being required to keep both.

Here's an example of where your merging strategy would produce incorrect results.

```json
{
  "type": "number",
  "$ref": "#/$defs/string",
  "$defs": {
    "string": { "type": "string" }
  }
```

If merged, the `"type": "number"` would override `"type": "string"` and string values would pass this schema. The correct behavior is that both assertions should apply and no values should pass this schemas because a value can't be a number and a string at the same time.

If the only keywords that conflict are annotations and your implementation only needs to use annotations in one particular way, then there's no harm in merging the way you have. It's not technically correct, but the end result is the same. However, you'll need to check to ensure that keywords that aren't annotations never get merged that way. Maybe you could merge annotations, but use the `allOf` wrapper for any conflicting non-annotations you encounter.

However, keep in mind that using reference merging as your sole means of disambiguating multiple annotations wouldn't be sufficient. You can have multiple annotations from a completely dereferenced schema. Take the schema above for example. It has no references, yet there are two `description` annotations that apply to the same location. Since you have to deal with multiple annotations in a more generic way anyway, I'm not sure there's a benefit to merging them with references.

**Comment by char0n** [comment]
Hi @Relequestual, thanks for the reactin.

> Removal of all references is not something supported by JSON Schema. There are some cases where it's not possible, and many cases where it will result in a change of behaviour and validation result.

> Any tooling that requires removal of references is unable to fully support JSON Schema 2019-09 or above.

> Why do you want to remove all references? Is it due to downstream tooling?
We added bundling to the specification to make handling external references easier, but you cannot and should not try to remove all references.

Yes, I'm aware of this, I've carefully read the https://json-schema.org/draft/2019-09/draft-handrews-json-schema-02#rfc.appendix.B.2. I'm not using dereferenced JSON Schema for later validation, but for other purposes like rendering, data generation, and others...where the representation is only approximation of the original intention not 100% reflection of it.

Let's say for the sake of the argument that I want to dereference the JSON Schema mentioned in https://github.com/orgs/json-schema-org/discussions/526#discussioncomment-7563198 (assuming it's Draft 2019-09 or 2020-12). I couldn't find one example of the dereferenced structure within the docs that would demonstrate how to do it, just the section (https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-00#rfc.section.7.7.1.1) that describes the result of dereferencing in words and then the reference of @jdesrosiers where he is [mentioning the `allOf` keyword](https://github.com/orgs/json-schema-org/discussions/526#discussioncomment-7559030).

I'm currently assuming that it would look like this:

**Dereference JSON Schema**:

```json
{
    "title": "Feature list",
    "type": "array",
    "prefixItems": [
        {
            "title": "Feature A",
            "properties": {
                "enabled": {
                   "allOf": [
                        {
                            "title": "Enabled",
                            "description": "generic description",
                            "type": ["boolean", "null"],
                            "default": null
                        },
                        { "default": true }
                   ] 
                }
            }
        },
        {
            "title": "Feature B",
            "properties": {
                "enabled": {
                    "allOf": [
                        {
                            "title": "Enabled",
                            "description": "generic description",
                            "type": ["boolean", "null"],
                            "default": null
                        },
                        { "description": "Feature B description" }
                   ] 
                }
            }
        }
    ],
    "$defs": {
        "enabledToggle": {
            "title": "Enabled",
            "description": "generic description",
            "type": ["boolean", "null"],
            "default": null
        }
    }
}
```

Is my assumption correct?

**Reply by Relequestual** [reply]
> I'm not using dereferenced JSON Schema for later validation, but for other purposes like rendering, data generation, and others...where the representation is only approximation of the original intenation not 100% reflection of it.

There exists data generation tooling that works with references. We should upgrade any tooling that doesn't support references, to support references. There is ongoing work to create libraries which handle all of the complexities so you don't have to.

> I couldn't find one example of the dereferenced structure within the docs that would demonstrate how to do it

This is because our recommendation is to not do this. In the case of dyanmic references, it may not be even possible.

> I'm currently assuming that it would look like this:
> Dereferenced JSON Schema:

Yes. But as mentioned, this will not always be possible. Cyclical references can't be fully removed, and in some cases dynamic references do not know the resolution location till runtime.

As an aside, we really appreciate you being interested and asking questions in our community. Keep it up =]

**Comment by char0n** [comment]
Hi @Relequestual and @jdesrosiers.

@jdesrosiers [confirmed](https://github.com/orgs/json-schema-org/discussions/526#discussioncomment-7568711) that the theoretical reference removal will have following form:

```json
"enabled": {
  "allOf": [{
    "title": "Enabled",
    "description": "generic description",
    "type": ["boolean", "null"],
    "default": null
  }],
  "description": "Feature B description",
}
```

I assumed in this [comment](https://github.com/orgs/json-schema-org/discussions/526#discussioncomment-7564673) that it will have the following form and @Relequestual confirmed it:

```json
"enabled": {
  "allOf": [
    {
      "title": "Enabled",
      "description": "generic description",
      "type": ["boolean", "null"],
      "default": null
    },
    {
      "description": "Feature B description"
    }
  ]
}
```

In terms of validation, both schema objects will behave similarly (more like equivalently?), as they both ultimately set `description` keyword to `Feature B description` by default if no other value is provided. However, their structure and the conceptual relationship between the `description` keyword and the `allOf` keyword differ.

Which one would be more correct form? Can we agree on one form?

**Reply by Relequestual** [reply]
In the simplest use case, the approach I showed would work, but not in all situations, hence @jdesrosiers's.

Consider the following two schemas, followed by a "dereferenced" version.
Because of `unevaluatedProperties`, you can't wrap the root level in another `allOf` item object, as it would change the results of validation.

```json
{
  "properties": {
    "foo": true,
    "bar": true
  },
  "$ref": "/schemaB.json",
  "unevaluatedProperties": false
}
```
```json
{
  "$id": "/schemaB.json",
  "properties": {
    "baz": true
  }
}
```

```json
{
  "allOf": [
    {
      "properties": {
        "foo": true,
        "bar": true
      },
      "unevaluatedProperties": false
    },
    {
      "properties": {
        "baz": true
      }
    }
  ]
}
```
```json
{
  "foo": 1,
  "bar": 2,
  "baz": 3
}
```

By the time you try to work out and cater for all the possible edge cases, it might have been simpler to just support processing references!

**Reply by jdesrosiers** [reply]
> Which one would be more correct form? Can we agree on one form?

@Relequestual's excellent explanation shows why the singleton `allOf` is the better choice, but ultimately, you're outside of what is allowed by the spec no matter what you choose. There is no "correct" form in this case.

**Comment by char0n** [comment]
Thanks both of you for patiently answering my question. Really appreciate it.
